HashLock, SecretLock, SecretProof, Mosiac Completion, General Updates

the-realtom opened the pull request:
## Changes
- Added HashLockTransaction
- Added SecretLockTransaction
- Added SecretProofTransaction
- Finished the remained of MosaicDefinitionTransaction and MosaicSupplyChangeTransaction (mostly just adding the layouts, and fixing up tests)
- Made all the currently implemented transaction types compatible as InnerTransaction (for when they are part of an Aggregate)
- Removed redundant imports/code

All unit and device tests are passing.

the-realtom committed:
Updated MosaicDefinition

Added layout, fixed test.

the-realtom committed:
Updated MosaicAlias

Fixed layout issue, moved unit test into namespace file.

the-realtom committed:
Added SecretProof

chrdevmar commented on core/src/apps/nem2/secret_lock/layout.py:
```diff
@@ -0,0 +1,118 @@
+from apps.common.layout import split_address
+
+enum_to_friendly_text = {
+    NEM2_SECRET_LOCK_SHA3_256: "SHA-3 266",
```
theres a typo here - 266 instead of 256

chrdevmar commented on core/tests/test_apps.nem2.hash_lock.py:
```diff
@@ -0,0 +1,36 @@
+
+@unittest.skipUnless(not utils.BITCOIN_ONLY, "altcoin")
+class TestNem2HashLock(unittest.TestCase):
+    # AggregateBonded
```
are these comments right here?

chrdevmar commented on core/tests/test_apps.nem2.hash_lock.py:
```diff
@@ -0,0 +1,36 @@
+                amount=10000000
+            ),
+            duration=480,
+            hash="4FF8C7E00B02D7380900B7795B4CDDC5B9D03536B5BD896BF54F8E5C53F4A32B"
```
oh so is the hash of an AggregateBonded transaction containing a transfer and a MosaicDefinition?

chrdevmar commented on core/src/apps/nem2/mosaic/layout.py:
```diff
@@ -1,80 +1,94 @@
+    elif mosaic_supply.action == NEM2_MOSAIC_SUPPLY_CHANGE_ACTION_INCREASE:
+        msg = "Increase supply by " + str(mosaic_supply.delta) + " whole units?"
+    else:
+        raise ValueError("Invalid supply change type")
```
this should be handled in the validator function for mosaic supply transactions

the-realtom committed:
Updated MosiacSupplyChange

Added layout, updated tests.

the-realtom committed:
Added SecretLock transaction type

the-realtom committed:
General cleanup

- Added all current transaction types to AggregateTransaction maps and proto
- Added EmbeddedTransactionCommon types where they didn't exist
- Cleaned up unused code and imports plus other general fixes
- Updated MosaicAlias validator and added validators for MosaicDefinition and MosaicSupply

the-realtom committed:
Added HashLockTransaction
